src/Raven.Server/Documents/Indexes/IndexStore.cs

```diff
@@ -57,40 +56,103 @@ public class IndexStore : IDisposable
 
         internal Action<(string IndexName, bool DidWork)> IndexBatchCompleted;
 
         private const int PathLengthLimit = 259; // Roslyn's MetadataWriter.PathLengthLimit = 259 
 
+
         internal static int MaxIndexNameLength = PathLengthLimit -
                                                  IndexCompiler.IndexNamePrefix.Length -
                                                  1 - // "."
                                                  36 - // new Guid()
                                                  IndexCompiler.IndexExtension.Length -
-                                                 4; // ".dll" 
-            
+                                                 4; // ".dll"
+
         public IndexStore(DocumentDatabase documentDatabase, ServerStore serverStore)
         {
             _documentDatabase = documentDatabase;
             _serverStore = serverStore;
-            _logger = LoggingSource.Instance.GetLogger<IndexStore>(_documentDatabase.Name);
+            Logger = LoggingSource.Instance.GetLogger<IndexStore>(_documentDatabase.Name);
 
             var stoppedConcurrentIndexBatches = _documentDatabase.Configuration.Indexing.NumberOfConcurrentStoppedBatchesIfRunningLowOnMemory;
             StoppedConcurrentIndexBatches = new SemaphoreSlim(stoppedConcurrentIndexBatches);
         }
 
-        public void HandleDatabaseRecordChange(DatabaseRecord record, long index)
+        public void HandleDatabaseRecordChange(DatabaseRecord record, long raftIndex)
         {
             if (record == null)
                 return;
 
-            HandleSorters(record, index);
-            HandleDeletes(record, index);
-            HandleChangesForStaticIndexes(record, index);
-            HandleChangesForAutoIndexes(record, index);
+            var indexesToStart = new List<Index>();
+
+            HandleSorters(record, raftIndex);
+            HandleDeletes(record, raftIndex);
+
+            HandleChangesForStaticIndexes(record, raftIndex, indexesToStart);
+            HandleChangesForAutoIndexes(record, raftIndex, indexesToStart);
+
+            if (indexesToStart.Count <= 0)
+                return;
+
+            var sp = Stopwatch.StartNew();
+
+            if (Logger.IsInfoEnabled)
+                Logger.Info($"Starting {indexesToStart.Count} indexes");
+
+            ExecuteForIndexes(indexesToStart, index =>
+            {
+                var indexLock = GetIndexLock(index.Name);
+
+                try
+                {
+                    indexLock.Wait(_documentDatabase.DatabaseShutdown);
+                }
+                catch (OperationCanceledException e)
+                {
+                    _documentDatabase.RachisLogIndexNotifications.NotifyListenersAbout(raftIndex, e);
+                    return;
+                }
+
+                try
+                {
+                    StartIndex(index);
+                }
+                catch (Exception e)
+                {
```

Comment:
Should we continue to try and create the next index? 
Since we will notify the client about a failure, but still will try to create the next index,

Reply:
This is running in parallel, the next index might have already run